autogpt_platform/backend/backend/integrations/embeddings.py

```diff
@@ -0,0 +1,163 @@
+        """
+        Generate embeddings for multiple texts (batch).
+
+        Args:
+            texts: List of texts to generate embeddings for.
+
+        Returns:
+            List of embedding vectors, one per input text.
+
+        Raises:
+            ValueError: If any text is invalid or batch size exceeds limit.
+            openai.APIError: If the OpenAI API call fails.
+        """
+        # Input validation
+        if not texts:
+            raise ValueError("Texts list cannot be empty")
+        if len(texts) > MAX_BATCH_SIZE:
+            raise ValueError(f"Batch size exceeds maximum of {MAX_BATCH_SIZE} texts")
+        for i, text in enumerate(texts):
+            if not text or not text.strip():
+                raise ValueError(f"Text at index {i} cannot be empty")
+            if len(text) > MAX_TEXT_LENGTH:
+                raise ValueError(
+                    f"Text at index {i} exceeds maximum length of {MAX_TEXT_LENGTH} characters"
+                )
+
+        try:
+            response = await self.client.embeddings.create(
+                model=EMBEDDING_MODEL,
+                input=texts,
+                dimensions=EMBEDDING_DIMENSIONS,
+            )
+            # Sort by index to ensure correct ordering
+            sorted_data = sorted(response.data, key=lambda x: x.index)
+            return [item.embedding for item in sorted_data]
+        except openai.APIError as e:
+            logger.error(f"OpenAI API error generating embeddings: {e}")
+            raise
+
+
+def create_search_text(name: str, sub_heading: str, description: str) -> str:
+    """
+    Combine fields into searchable text for embedding.
+
+    This creates a single text string from the agent's name, sub-heading,
+    and description, which is then converted to an embedding vector.
+
+    Args:
+        name: The agent name.
+        sub_heading: The agent sub-heading/tagline.
+        description: The agent description.
+
+    Returns:
+        A single string combining all non-empty fields.
+    """
+    parts = [name or "", sub_heading or "", description or ""]
+    return " ".join(filter(None, parts)).strip()
+
+
+# Singleton instance with lock for thread-safe initialization
+_embedding_service: Optional[EmbeddingService] = None
+_embedding_service_lock: asyncio.Lock = asyncio.Lock()
+
+
+async def get_embedding_service() -> EmbeddingService:
+    """
+    Get or create the embedding service singleton.
+
+    Uses double-checked locking to prevent race conditions in concurrent
+    async environments while avoiding lock overhead after initialization.
+
+    Returns:
+        The shared EmbeddingService instance.
+
+    Raises:
+        ValueError: If OpenAI API key is not configured.
+    """
+    global _embedding_service
+    if _embedding_service is None:
+        async with _embedding_service_lock:
+            if _embedding_service is None:
+                _embedding_service = EmbeddingService()
```

Comment:
We can make this return embeddingservice(); and annotate the function with cachedproperty

No lock needed. Accessed like a field. No global. Lazily evaluated.